### PR TITLE
implement support for new relic

### LINKED ISSUE #258: Implement Support for New Relic
Add support for structured logging and instrumenting http requests via new relic

### PR DESCRIPTION
**Pull request by Dotunj**
This adds support for instrumenting http requests via new relic. This resolves #258 

Also added support for a logger interface and a corresponding implementation that simply writes to standard output. This also provides support for optional logging of http requests.

### CODE REVIEW

File: server/middleware.go [hunk 1]
```diff
@@ -455,6 +482,139 @@ func pagination(next http.Handler) http.Handler {
...
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			cfg, err := config.Get()
+			if err != nil {
+				return
+			}
+
+			ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+
+			if isLogHttpRequestEnabled(cfg.Logger) && cfg.Logger.Type != "" {
```
Comment (Subomi) [inline server/middleware.go:496]: I don't think we need to check if the Logger type is not nil here. The system should default to console logger where the user don't specify a different logger.

File: logger/console.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package logger
```
Comment (Subomi) [inline logger/console.go:1]: Let's write test for this guy.

File: cmd/main.go [hunk 3]
```diff
@@ -98,6 +102,20 @@ func main() {
 				}
 			}
 
+			if cfg.Logger.Type != "" {
+				lo, err = logger.NewLogger(cfg.Logger)
```
Comment (Subomi) [inline cmd/main.go:106]: Where a logger is not provided, the system should default to console logger with log level -- info.

File: config/config.go [hunk 4]
```diff
@@ -84,6 +84,22 @@ type GroupConfig struct {
...
 
+type LoggerConfiguration struct {
+	Type  LoggerProvider `json:"type"`
+	Level string         `json:"level"`
+	LogHttpRequest bool  `json:"log_http_request"`
+}
+
+type TracerConfiguration struct {
+	Type     TracerProvider `json:"type"`
+	NewRelic struct {
```
Comment (Subomi) [inline config/config.go:95]: NewRelic configuration should be standalone -- `NewRelicConfig`

File: tracer/tracer.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package tracer
```
Comment (Subomi) [inline tracer/tracer.go:1]: I'm not sure we should be doing tracing yet. I'd prefer us to figure out logs & metrics first.

**Review by Subomi — CHANGES_REQUESTED** [review]
@Dotunj FYI boss.

File: tracer/tracer.go (hunk 5 above)
Comment (Dotunj) [reply tracer/tracer.go:1]: If we're not doing tracing yet, then there might be no need for New Relic at the moment.

Comment (Subomi) [reply tracer/tracer.go:1]: I don't think so -- Error tracking, Instrumentation.

File: config/config.go [hunk 6]
```diff
@@ -84,6 +84,22 @@ type GroupConfig struct {
 	DisableEndpoint bool                   `json:"disable_endpoint"`
 }
 
+type LoggerConfiguration struct {
+	Type  LoggerProvider `json:"type"`
+	Level string         `json:"level"`
+	LogHttpRequest bool  `json:"log_http_request"`
```
[the remark this answers is not in the document]
Comment (Subomi) [reply config/config.go:90]: We can instead tie the http requests to a particular log level vs. creating an new config level.

[the remark this answers is not in the document]
Comment (Dotunj) [reply config/config.go:90]: what log level are we going with for this?

[the remark this answers is not in the document]
Comment (Subomi) [reply config/config.go:90]: Info.

**Review by Subomi — APPROVED** [review]

**Review by danvixent — APPROVED** [review]

**Review by jirevwe — APPROVED** [review]

**Pull request merged**